### PR TITLE
 [CI] Add some basic end-to-end tests

### LINKED ISSUE #94: [CI] Add some basic end-to-end tests
**Current Behavior**

Currently, there are no automatic tests that help make refactoring safe(r). Testing has to be performed manually. 

**Desired Behavior**

A basic set of simple end-to-end tests that check basic functionality. Initially, the tests are intended to be used only locally, i.e. not as part of the CI pipeline. At first, the tests will not be fully automated, i.e. require some manual setup, e.g. setting up a local cluster. 
 
---
**Resources**

* https://github.com/bats-core/bats-core - Bats is a TAP-compliant testing framework for Bash. It provides a simple way to verify that the UNIX programs you write behave as expected.
* https://github.com/fullstorydev/grpcurl - grpcurl is a command-line tool that lets you interact with gRPC servers. It's basically curl for gRPC servers.

**Alternatives / Additional Context**

The basic pattern of these tests is to install the mesh in the cluster, wait until all rollouts are ready, install sample applications, wait until all rollouts are ready, possibly test the sample application, delete the sample application, repeat for all sample applications, and finally delete the mesh.

### PR DESCRIPTION
**Pull request by mgfeller**
**Description**

This PR fixes #94 

**Notes for Reviewers**

Instructions on how to set up and run the tests can be found in `tests/README.md`. 

Expected output can be found here: https://github.com/layer5io/meshery-consul/issues/94#issuecomment-692199238

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**

- [x] Yes, I signed my commits.

### CODE REVIEW

**Review by kushthedude — COMMENTED** [review]
I would prefer regular bash scripts to be instantiated in the CI rather than having BATS.

**Comment by mgfeller** [comment]
@kushthedude I'm not quite sure I understand what you mean. Would you prefer to replace BATS with something else? 

IMHO BATS is a very convenient framework for testing from the shell leveraging existing command line tools. Currently, these tests are not intended for the CI workflows on GitHub yet.

**Comment by kushthedude** [comment]
Thats what I am trying to say , replace bats with bash script which could
be invoked in CI.

**Comment by mgfeller** [comment]
Ok, why? BATS scripts can also be invoked in CI.

**Comment by kushthedude** [comment]
Because CI has its own way of asserting and checking, we can define
seperate jobs and assertions like done in BATS. Moreover, bats doesnt have
native support in GitHub action.BATS could have been beneficial if we had
travis.

But yes the following PR adds capability to run tests on local, but we will
have to amend it when having it run on action runner. If others agree to
have both, then let's go ahead.

**Comment by mgfeller** [comment]
Ok, I see, thanks for the explanation :+1: Do we have tests like this that run on CI for another adapter? So far, in the consul adapter, there are no tests whatsoever. For my part, I want to have tests like this that I can also run locally easily, so that I can test early and fast :-)

**Review by Aisuko — COMMENTED** [review]
Here is a little suggestions below

**Comment by Aisuko** [comment]
In my opinion, we can take both of them, run `bat` at local , add `bash` into the CI scripts for CI situation. Sometime when we development at local, we  need to do some local test

File: main.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func main() {
 	s := grpc.NewServer(
 	// grpc.Creds(credentials.NewServerTLSFromCert(&insecure.Cert)),
 	)
+	reflection.Register(s)
```
[the remark this answers is not in the document]
Comment (mgfeller) [reply main.go:58]: Yes, that's a good point :+1:

**Review by Aisuko — APPROVED** [review]
approve

**Comment by mgfeller** [comment]
@leecalcote I wonder whether you could have a look at this PR, and possibly approve and merge it? Thanks in advance!

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+* A local cluster, e.g. microk8s. The tests were tested using microk8s 1.16/stable.
+* `kubectl`
+* `jq`, a tool for processing JSON inputs.
+* BATS, the Bash Automated Testing System: https://github.com/bats-core/bats-core
+* gRPCurl, a command-line tool to interact with gRPC servers: https://github.com/fullstorydev/grpcurl
+* The environment variable `MESHERY_ADAPTER_ADDR` is defined and specifies the IP address or DNS name of the meshery
+ adapter, e.g. `localhost`. For some configurations, this might be set up in the Makefile.
+* The adapter is running.
+
```
Comment (leecalcote) [inline tests/README.md:23]: We might be missing `base64` as a prerequisite.

Comment (mgfeller) [reply tests/README.md:23]: I'll add that.

**Review by leecalcote — APPROVED** [review]
Output from a clean Docker Desktop system:

```
➜  meshery-consul git:(feature/mgfeller/94) make e2e-tests-localhost
Running end-to-end tests on 'localhost'
cd tests && bats e2e/
 ✓ gRPCurl should be installed
 ✓ jq should be installed
 ✓ base64 should be installed
    current kubectl context is: docker-desktop                                                4/34
 ✓ kubectl should be installed
 ✓ cluster should be reachable
 ✓ all pods should be ready
 ✓ adapter should return expected supported operations
 ✓ adapter should return expected mesh name
 ✓ client instance should be created
 ✗ consul_install should be successful
   (in test file e2e/ca_install_mesh_tests.bats, line 34)
     `[ "$status" -eq 0 ]' failed
 ✗ deployment/consul-consul-connect-injector-webhook-deployment should be ready
   (in test file e2e/ca_install_mesh_tests.bats, line 39)
     `[ "$status" -eq 0 ]' failed
 ^Cmake: *** [e2e-tests-localhost] Terminated: 15dy                                          12/34
[1]    73930 terminated  make e2e-tests-localhost
➜  meshery-consul git:(feature/mgfeller/94) clear
➜  meshery-consul git:(feature/mgfeller/94) make e2e-tests-localhost
Running end-to-end tests on 'localhost'
cd tests && bats e2e/
 ✓ gRPCurl should be installed
 ✓ jq should be installed
 ✓ base64 should be installed
    current kubectl context is: docker-desktop                                                4/34
 ✓ kubectl should be installed
 ✓ cluster should be reachable
 ✓ all pods should be ready
 ✓ adapter should return expected supported operations
 ✓ adapter should return expected mesh name
 ✓ client instance should be created
 ✓ consul_install should be successful
 ✓ deployment/consul-consul-connect-injector-webhook-deployment should be ready
 ✓ statefulset/consul-consul-server should be ready
 ✓ daemonset/consul-consul should be ready
 ✓ httpbin installation should be successful
 ✓ deployment/httpbin should be ready
 ✓ httpbin deletion should be successful
 ✓ no resources should remain in the httpbin namespace
 ✓ bookinfo installation should be successful
 ✓ deployment/details-v1 should be ready
 ✓ deployment/ratings-v1 should be ready
 ✓ deployment/reviews-v1 should be ready
 ✓ deployment/reviews-v2 should be ready
 ✓ deployment/reviews-v3 should be ready
 ✓ deployment/productpage-v1 should be ready
 ✓ bookinfo deletion should be successful
 ✓ no resources should remain in the bookinfo namespace
 ✓ image-hub installation should be successful
 ✓ deployment/web-deployment should be ready
 ✓ deployment/api-deployment-v1 should be ready
 ✓ deployment/envoy should be ready
 ✓ image-hub deletion should be successful
 ✓ no resources should remain in the image-hub namespace
 ✓ consul_install delete should be successful
 ✓ no resources should remain in the consul namespace

34 tests, 0 failures
```

**Pull request merged**